packages/search-input/src/SearchInput/SearchInput.tsx

```diff
@@ -0,0 +1,433 @@
+
+import {
+  baseInputStyle,
+  clearButtonSizeStyle,
+  formStyle,
+  inputThemeStyle,
+  inputWrapperDisabledStyle,
+  inputWrapperFocusStyles,
+  inputWrapperInteractiveStyles,
+  inputWrapperSizeStyle,
+  inputWrapperStyle,
+  inputWrapperThemeStyle,
+  searchIconDisabledStyle,
+  searchIconSizeStyle,
+  searchIconThemeStyle,
+} from './SearchInput.styles';
+import { SearchInputProps, Size, State } from './SearchInput.types';
+
+/**
+ * # SearchInput
+ *
+ * SearchInput component
+ *
+ * @param props.state The current state of the SearchInput. This can be none, or loading.
+ * @param props.darkMode determines whether or not the component appears in dark theme.
+ * @param props.size determines the size of the text and the height of the input.
+ * @param props.baseFontSize determines the base font size if size is set to default.
+ */
+
+export const SearchInput = React.forwardRef<HTMLInputElement, SearchInputProps>(
+  function SearchInput(
+    {
+      placeholder = 'Search',
+      className,
+      darkMode: darkModeProp,
+      size = Size.Default,
+      disabled,
+      children,
+      state = State.Unset,
+      value: valueProp,
+      onChange: onChangeProp,
+      onSubmit: onSubmitProp,
+      'aria-label': ariaLabel,
+      'aria-labelledby': ariaLabelledBy,
+      ...rest
+    }: SearchInputProps,
+    forwardRef: React.Ref<HTMLInputElement>,
+  ) {
+    const { theme, darkMode } = useDarkMode(darkModeProp);
+    const [isOpen, setOpen] = useState(false);
+    // The index of the currently highlighted result option
+    const [highlightIndex, setHighlightIndex] = useState<number>(0);
+
+    const closeMenu = () => setOpen(false);
+    const openMenu = () => setOpen(true);
+
+    const formRef = useRef<HTMLFormElement>(null);
+    const searchBoxRef = useRef<HTMLDivElement>(null);
+    const menuRef = useRef<HTMLUListElement>(null);
+    const inputRef = useForwardedRef(forwardRef, null);
+    const clearButtonRef = useRef<HTMLButtonElement>(null);
+    const resultRefs = useDynamicRefs<HTMLElement>({ prefix: 'result' });
+    const [focusedElement, trackFocusedElement] = useState<Element>();
+    const highlightedElementRef = resultRefs(`${highlightIndex}`);
+
+    const withTypeAhead = !isUndefined(children);
+
+    const { value, handleChange } = useControlledValue(valueProp, onChangeProp);
+
+    /** Fires a change event to update the input value */
+    const changeInputValue = useCallback(
+      (newVal: string) => {
+        if (inputRef.current) {
+          // We change the element's value
+          // and then make sure that the change event is fired
+          inputRef.current.value = newVal;
+          const nativeChangeEvent = new Event('change', {
+            cancelable: true,
+            bubbles: true,
+          });
+          const syntheticChangeEvent = createSyntheticEvent(
+            nativeChangeEvent,
+            inputRef.current,
+          );
+          handleChange(syntheticChangeEvent);
```

Comment:
why do this instead of just using the event itself?

Reply:
Since `changeInputValue` can be called programatically we need to trigger the `change` event with an event. The handler expects a `React.SyntheticEvent`, and not a native event.